File: web/src/screens/splashScreen.ts
import { rgb } from '../utils.js';
import { Screen } from '../screen.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';

export class SplashScreen extends Screen {
  private frameCounter = 0;
  private readonly logo: Animation;

  constructor() {
    super();
    this.logo = Resource.getAnimation('data/images/logo.bmp', 1);
  }

  onDraw(ctx: CanvasRenderingContext2D): void {
    let yOffset = -160 + this.frameCounter;
    if (yOffset === 0) Sound.playSample('data/sounds/boot');
    if (yOffset > 0) yOffset = 0;

    ctx.fillStyle = rgb(231, 215, 156);
    ctx.fillRect(0, 0, 320, 240);
    this.logo.drawFrame(ctx, 0,
      160 - this.logo.getFrameWidth() / 2,
      120 - this.logo.getFrameHeight() / 2 + yOffset);
  }

  onLogic(): void {
    if (this.frameCounter > 250) {
      this.exit();
      return;
    }
    this.frameCounter++;
  }
}

File: wii/src/screens/splash_screen.cpp
#include "screens/splash_screen.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "constants.h"

SplashScreen::SplashScreen() {
    logo_ = Resource::getAnimation("data/images/logo.bmp", 1);
}

void SplashScreen::onDraw(SDL_Renderer* renderer) {
    int yOffset = -160 + frameCounter_;
    if (yOffset == 0) Sound::playSample("data/sounds/boot");
    if (yOffset > 0) yOffset = 0;

    SDL_SetRenderDrawColor(renderer, 231, 215, 156, 255);
    SDL_Rect bg = {0, 0, SCREEN_WIDTH, SCREEN_HEIGHT};
    SDL_RenderFillRect(renderer, &bg);

    logo_->drawFrame(renderer, 0,
        160 - logo_->getFrameWidth() / 2,
        120 - logo_->getFrameHeight() / 2 + yOffset);
}

void SplashScreen::onLogic() {
    if (frameCounter_ > 250) {
        exit();
        return;
    }
    frameCounter_++;
}
